include/zim/suggestion.h

```diff
@@ -0,0 +1,190 @@
+ * is provided AS IS, WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, and
+ * NON-INFRINGEMENT.  See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ */
+
+#ifndef ZIM_SUGGESTION_H
+#define ZIM_SUGGESTION_H
+
+#include "suggestion_iterator.h"
+#include "archive.h"
+
+namespace Xapian {
+  class Enquire;
+  class MSet;
+};
+
+namespace zim
+{
+
+class SuggestionSearcher;
+class SuggestionSearch;
+class SuggestionIterator;
+class SuggestionDataBase;
+class SuggestionQuery;
+
+/**
+ * A SuggestionSearcher is a object suggesting over titles of an Archive
+ *
+ * A SuggestionSearcher is mainly used to create new `SuggestionSearch`
+ * Internaly, this is a wrapper around a SuggestionDataBase with may or may not
+ * include a Xapian index.
+ *
+ * You should consider that all search operations are NOT threadsafe.
+ * It is up to you to protect your calls to avoid race competition.
+ * However, SuggestionSearcher (and subsequent classes) do not maintain a global/
+ * share state You can create several Searchers and use them in different threads.
+ */
+class SuggestionSearcher
+{
+  public:
+    /** SuggestionSearcher constructor.
+     *
+     * Construct a SuggestionSearcher on top of an archive.
+     *
+     * @param archive An archive to suggest on.
+     */
+    explicit SuggestionSearcher(const Archive& archive);
+
+    SuggestionSearcher(const SuggestionSearcher& other);
+    SuggestionSearcher& operator=(const SuggestionSearcher& other);
+    SuggestionSearcher(SuggestionSearcher&& other);
+    SuggestionSearcher& operator=(SuggestionSearcher&& other);
+    ~SuggestionSearcher();
+
+    /** Create a SuggestionSearch for a specific query.
+     *
+     * The search is made on the archive under the SuggestionSearcher.
+     *
+     * @param query The SuggestionQuery to search.
+     */
+    SuggestionSearch suggest(const SuggestionQuery& query);
+
+    /** Set the verbosity of search operations.
+     *
+     * @param verbose The verbose mode to set
+     */
+    void setVerbose(bool verbose);
+
+  private: // methods
+    void initDatabase();
+
+  private: // data
+    std::shared_ptr<SuggestionDataBase> mp_internalDb;
+    Archive m_archive;
+};
+
+/**
+ * A SuggestionQuery represent a query.
+ *
+ * It describe what have to be searched for suggestion.
+ */
+class SuggestionQuery
+{
+  public:
+    /** SuggestionQuery constructor.
+     *
+     * Create a empty query.
+     */
+    SuggestionQuery() = default;
+
+    /** Set the textual query of the SuggestionQuery.
+     *
+     * @param query The string to search for.
+     */
+    SuggestionQuery& setQuery(const std::string& query);
+
+    std::string m_query { "" };
+};
```

Comment:
Now that `SuggestionQuery` contains only a string, we can simply remove it and make `SuggestionSearcher::suggest` take a string as argument.

Reply:
That makes sense. Done!